File: frontend/src/types/workflow.ts
import { Node, Edge } from 'reactflow'

export type NodeStatus = 'idle' | 'thinking' | 'executing' | 'completed' | 'error' | 'interrupted'

export interface AgentNodeData {
  label: string
  type: 'llm' | 'tool' | 'router' | 'map' | 'research' | 'writer' | 'reviewer'
  status: NodeStatus
  modelConfig?: {
    provider: 'openai' | 'anthropic'
    model: string
    temperature?: number
    maxTokens?: number
    // RAG configuration
    enableRag?: boolean
    knowledgeDocuments?: string[]
    ragMode?: 'document' | 'chunk'
    ragTopK?: number
    ragMinScore?: number
  }
  prompt?: string
  streamingContent?: string
  error?: string
  toolConfig?: {
    toolName: string
    args: Record<string, any>
  }
  // RAG search results (from WebSocket)
  ragSearchResults?: Array<{
    documentId: string
    title: string
    score: number
    snippet?: string
  }>
}

export interface CanvasViewport {
  x: number
  y: number
  zoom: number
}

export interface WorkflowCanvas {
  nodes: Node<AgentNodeData>[]
  edges: Edge[]
  viewport: CanvasViewport
}

export interface Workflow {
  id: string
  ownerId: string
  title: string
  description: string
  canvasJson: WorkflowCanvas
  sourceTemplateId?: string
  isPublic: boolean
  createdAt: string
  updatedAt: string
}

export interface WorkflowRun {
  id: string
  workflowId: string
  threadId: string
  status: 'running' | 'completed' | 'failed' | 'interrupted'
  inputSummary: string
  startedAt: string
  finishedAt?: string
}

export interface WorkflowCreateInput {
  title: string
  description: string
  canvasJson?: WorkflowCanvas
  sourceTemplateId?: string
}

export interface WorkflowUpdateInput {
  title?: string
  description?: string
  canvasJson?: WorkflowCanvas
  isPublic?: boolean
}

File: backend/app/models/workflow.py
from sqlalchemy import Column, String, Text, Boolean, DateTime, ForeignKey, Enum
from sqlalchemy.dialects.postgresql import UUID, JSONB
from datetime import datetime
import uuid
import enum

from app.core.database import Base

class Workflow(Base):
    __tablename__ = "workflows"
    
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    owner_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    title = Column(String, nullable=False)
    description = Column(Text, nullable=True)
    canvas_json = Column(JSONB, nullable=False, default=dict)
    source_template_id = Column(UUID(as_uuid=True), nullable=True)
    is_public = Column(Boolean, default=False, nullable=False)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)

class RunStatus(str, enum.Enum):
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    INTERRUPTED = "interrupted"

class WorkflowRun(Base):
    __tablename__ = "workflow_runs"
    
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    workflow_id = Column(UUID(as_uuid=True), ForeignKey("workflows.id"), nullable=False)
    thread_id = Column(UUID(as_uuid=True), nullable=False, unique=True)
    status = Column(Enum(RunStatus), default=RunStatus.RUNNING, nullable=False)
    input_summary = Column(Text, nullable=True)
    started_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    finished_at = Column(DateTime, nullable=True)
